dev-docs/metrics.md

```diff
@@ -179,3 +179,98 @@ Strong references to instruments will be discarded when changes to the aggregate
 On the fast path the mutex is locked for the period that it takes to upgrade the weak reference. This is a fast operation, and should not block the thread for any meaningful period of time.
 
 If there is shown to be contention in future profiling we can revisit.
+
+## Adding new metrics
+There are different types of metrics.
+
+* Static - Used by us to monitor feature usage.
+* Dynamic - Defined by the user to monitor the health and performance of their system.
+
+> New features should add BOTH static and dynamic metrics.
+
+> Why are static metrics no longer recommended for users to use directly?
+> 
+> They can, but usually it'll be only a starting point for them. We can't predict the things that users will want to monitor, and if we tried we would blow up the cardinality of our metrics resulting in high costs for our users via their APMs.
+> 
+> This *is* a departure from the previous way of doing things, but as dynamic metrics are not available we need to move away from statically defined metrics.
+
+### Naming
+Metrics should be named in a way that is consistent with the rest of the metrics in the system.
+
+**Metrics**
+* `<feature>` - This should be a noun that describes the feature that the metric is monitoring.
+
+* `<feature>.<verb>` - Sub-metrics are usually a verb that describes the action that the metric is monitoring.
+
+**Attributes**
+* `<feature>.<feature-specific-attribute>` - Are always prefixed with the feature name unless they are standard metrics from the [otel semantic conventions](https://opentelemetry.io/docs/specs/semconv/general/metrics/). 
+
+### Static metrics
+When adding a new feature to the Router you must also add new static metrics to monitor the usage of that feature and users cannot turn them off.
+These metrics must be low cardinality and not leak any sensitive information. Users cannot change these metrics and they are primarily for us to see how our features are used so that we can inform future development.
```

Comment:
a lot of static metrics actually monitor standard router operations and are not for us to collect data, but for users to observe the router.
If we want this to be the defining point, let's maybe not call them `static` VS `dynamic` metrics, but `internal` VS `monitoring` or `user` metrics, something like that?
I'd prefer we keep the distinction between `static` metrics as defined directly with `tracing`, and `dynamic` metrics as the ones defined by custom instruments, that can be activated with runtime conditions, and have another clear separation beween the metrics used for internal reporting (as with the `apollo.router.operations` and `apollo.router.config` prefixes) and the user facing ones.

Reply:
This is pretty much what I was trying to say:
* Static metrics defined by macro (not `tracing` as we are going to remove this), 
* Dynamic metrics configurable by yaml.

I think the rub though is that I'm trying to define where static metrics are best used.

The problem with static metrics is that attributes that we think are OK are not, and attributes that we don't include are required. For instance:

1. Feature X needs some metrics. We put some static metrics in using a macro.
2. User A: complains that they need operation name on the metric to identify which operations are causing an issue.
3. We add `graphql.operation.name` to a static metric.
4. User B: complains that they have now had their APM cardinality budget blown because they have unrestricted operation names and therefore they are high cardinality.

Ideal static metrics uses:
* Router lifecycle stuff.
* Global error rates.
* Feature usage.
* Stuff that happens outside of a request/response. Cache clears

Ideal dynamic metrics uses:
* Anything that happens during a request response.

It's not so much a question of internal vs external, so I'll change the text to reflect the above and remove: `Users cannot change these metrics and they are primarily for us to see how our features are used so that we can inform future development.`

Reply:
d233cf3bdf553f308d2702f47f3c79312ec0bfdc